File: LintCode_Python/singleNumberII.py
# Given 3*n + 1 numbers, every numbers occurs triple times except one, find it.

# Have you met this question in a real interview? Yes
# Example
# Given [1,1,2,3,3,3,2,2,4,1] return 4

# Challenge
# One-pass, constant extra space.


def singleNumberII(A):
    if not A:
        return None
    bits = [0 for i in range(32)]
    result = 0
    for i in range(len(bits)):             # i stand for bit
        for j in range(len(A)):            # j stand for number
            bits[i] += (A[j] >> i) & 1     # add the ith bits
        bits[i] %= 3                       # bits[i]=bits[i]%3  if the ith bits appears three times, bits[i]=0
        result |= bits[i] << i             # change the single number from binary to decimal
    return result

print singleNumberII([1, 1, 2, 3, 3, 3, 2, 2, 4, 1])

File: LeetCode_Java/src/SingleNumberII.java
/**
 * Given an array of integers, every element appears three times except for one. Find that single one.
 * Note:
 * Your algorithm should have a linear runtime complexity. Could you implement it without using extra memory?
 * 
 * Tags: Bit Manipulation
 * @author wendi
 *
 */
public class SingleNumberII {
	
	/**
	 * Method2: Bit Manipulation using ^
	 * 
	 * @param int[] nums
	 * @return int
	 * Time: O(32n)
	 * Space: O(32)
	 */
	public int singleNumberIII(int[] nums) {
		if (nums == null|| nums.length == 0) {
			return -1;
		}
		int result = 0;
		return result;
	}
	
	
	/**
	 * Method1: Bit Manipulation: the number in 32 bits and just count how many 1s are there in each 
	 * bit, and sum %= 3 will clear it once it reaches 3. After running for all the numbers for each 
	 * bit, if we have a 1, then that 1 belongs to the single number, we can simply move it back to 
	 * its spot by doing ans |= sum << i;
	 * @param int[] nums
	 * @return int
	 * Time: O(32n)
	 * Space: O(32)
	 */
	public int singleNumberII(int[] nums) {
		if (nums == null|| nums.length == 0) {
			return -1;
		}
		int result = 0;
		for (int i = 0; i < 32; i++) {
			int sum = 0;
			for (int num: nums) {
				sum += num >> i & 1;
			}
			result |= sum % 3 << i;
		}
		return result;
	}

	public static void main(String[] args) {
		// TODO Auto-generated method stub
		SingleNumberII result = new SingleNumberII();
		System.out.println(result.singleNumberII(new int[] {2, 3, 1, 2, 1, 2, 1}));
		System.out.println(result.singleNumberIII(new int[] {2, 3, 1, 2, 1, 2, 1}));
	}

}
